### PR TITLE
Zoltan2:  total UVM hack for empire

### PR DESCRIPTION
**Pull request by kddevin**
@trilinos/zoltan2 @srajama1 @jjellio @bathmatt 

## Description
Added a HACK that copies data returned by Tpetra::Map::getNodeElementList() from the UVM memory returned by Tpetra to host memory before sending the memory to Tpetra::Distributor.

## Motivation and Context
SpectrumMPI is reported to behave badly with UVM unless certain runtime flags are used, but using the flags slows down all MPI communication.  So this HACK avoids passing UVM to MPI for the one use case reported by Empire debuggers.
See
https://gitlab-ex.sandia.gov/jhu/TrilinosSolverPerformance/issues/43

I am not excited about merging this HACK.  The best solution is to obtain a working MPI.  The next best solution may be for Tpetra to identify UVM memory in Distributor and do the copy there for all users.  I hope to see whether this change benefits the debugging effort before actually merging it.

## How Has This Been Tested?

Built/tested on cee-lan and on white testbed with and without CUDA.
New test multiJaggedPremigrateTest.xml explicitly invokes pre-migration, including function mj_premigrate_to_subset which was called out by Empire debuggers.
@jjellio said he could test these changes in the application from a branch.

## Checklist

- [ ] My commit messages mention the appropriate GitHub issue numbers.
- [ ] My code follows the code style of the affected package(s).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [code contribution guidelines](../blob/master/CONTRIBUTING.md) for this project.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] No new compiler warnings were introduced.
- [ ] These changes break backwards compatibility.

### CODE REVIEW

**Comment by jjellio** [comment]
I can add it to my build.  I've got a patch from Matt, that we believe should eliminate UVM+MPI use in EMPIRE as well. So, with this we should be be able to assert that we aren't doing anything wrong w.r.t. ATS2 platforms.

**Comment by kddevin** [comment]
@jjellio Do you want this HACK merged with the develop branch?  Or do you want to test it from this branch?

**Comment by jjellio** [comment]
Do you want this hack in Trilinos? Or would you rather leave it out and wait for the MJ refactor?

I don't think this is the root of EMPIRE's issue. I just like eliminating all possible cases (we got the EMPIRE side clean)

I personally don't like the 'hacks' going into the code base, since they get forgotten and end up biting us later, but on the flip side it is annoying to use patched code bases rather than 'master/develop'.

**Comment by jjellio** [comment]
FYI, I'll test it using a patch from the PR. My Trilinos SHA is frozen until we resolve the issue at hand.

**Comment by mhoemmen** [comment]
I'm a bit confused.  `Tpetra::Map::getNodeElementList` returns `HostSpace` memory, not UVM memory:

https://github.com/trilinos/Trilinos/blob/c6816e23ff9c44a415bfaebee441e968085fc896/packages/tpetra/core/src/Tpetra_Map_decl.hpp#L1295
https://github.com/trilinos/Trilinos/blob/c6816e23ff9c44a415bfaebee441e968085fc896/packages/tpetra/core/src/Tpetra_Map_def.hpp#L1660

That was a fairly recent addition, so perhaps all the debugging is happening with some earlier version of the code that uses UVM memory.

File: packages/zoltan2/src/algorithms/partition/Zoltan2_AlgMultiJagged.hpp [hunk 1]
```diff
@@ -6783,7 +6783,37 @@ bool Zoltan2_AlgMJ<Adapter>::mj_premigrate_to_subset( int used_num_ranks,
...
+    // data to be sent.
     ArrayView<const mj_gno_t> sent_gnos(initial_mj_gnos_, num_local_coords_);
+#else
+    // Unfortunately, this HACK is needed to workaround BUGs in SpectrumMPI.
+    // Apparently, SpectrumMPI does not communicate UVM memory correctly 
+    // without special command-line flags that dramatically slow MPI 
+    // communications.  
+    // When Tpetra is used in the XpetraMultiVectorAdapter, 
+    // initial_mj_gnos_ is obtained from Tpetra::Map::getNodeElementList().
+    // For CUDA builds, Tpetra returns UVM memory in getNodeNumElements(),
```
Comment (mhoemmen) [inline packages/zoltan2/src/algorithms/partition/Zoltan2_AlgMultiJagged.hpp]: See my comments.  `Tpetra::Map::getNodeElementList` returns HostSpace memory, at least in Trilinos:master.

**Comment by jjellio** [comment]
Matt has the most recent run (I am not sure what SHA he is using)

It wasn't until the prior weekend that I wrote a stacktrace tool that got line numbers correct.

Before that, I could only point out the function name.

```
[36] (0x200000172b8b) MPI_Send src/generated_wrappers.code:22430
[35] (0x20004ab6d077) void Teuchos::(anonymous namespace)::sendImpl<long long>(long long const*, int, int, int, Teuchos::Comm<int> const&) ../packages/teuchos/comm/src/Teuchos_CommHelpers.cpp:705
[34] (0x20004ab2c2d7) void Teuchos::send<int, long long>(long long const*, int, int, int, Teuchos::Comm<int> const&) ../packages/teuchos/comm/src/Teuchos_CommHelpers.cpp:1194
[33] (0x2000280d9517) void Tpetra::Distributor::doPosts<long long>(Teuchos::ArrayRCP<long long const> const&, unsigned long, Teuchos::ArrayRCP<long long> const&) ../packages/tpetra/core/src/Tpetra_Distributor.hpp:1360
[32] (0x2000280da683) void Tpetra::Distributor::doPostsAndWaits<long long>(Teuchos::ArrayView<long long const> const&, unsigned long, Teuchos::ArrayView<long long> const&) ../packages/tpetra/core/src/Tpetra_Distributor.hpp:1076
[31] (0x2000280e4003) Zoltan2::Zoltan2_AlgMJ<Zoltan2::XpetraMultiVectorAdapter<Xpetra::MultiVector<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> > > >::mj_premigrate_to_subset(int, int, Teuchos::RCP<Zoltan2::Environment const>, Teuchos::RCP<Teuchos::Comm<int> const>, int, int, long long, unsigned long, long long const*, double**, int, double**, Teuchos::RCP<Teuchos::Comm<int> const>&, int&, long long*&, double**&, double**&, int*&) ../packages/zoltan2/src/algorithms/partition/Zoltan2_AlgMultiJagged.hpp:6787
[30] (0x2000280f4223) Zoltan2::Zoltan2_AlgMJ<Zoltan2::XpetraMultiVectorAdapter<Xpetra::MultiVector<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> > > >::partition(Teuchos::RCP<Zoltan2::PartitioningSolution<Zoltan2::XpetraMultiVectorAdapter<Xpetra::MultiVector<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> > > > > const&) ../packages/zoltan2/src/algorithms/partition/Zoltan2_AlgMultiJagged.hpp:6911
[29] (0x20002809ced7) Zoltan2::PartitioningProblem<Zoltan2::XpetraMultiVectorAdapter<Xpetra::MultiVector<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> > > >::solve(bool) ../packages/zoltan2/src/problems/Zoltan2_PartitioningProblem.hpp:633
[28] (0x200028117587) MueLu::Zoltan2Interface<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> >::Build(MueLu::Level&) const ../packages/muelu/src/Rebalancing/MueLu_Zoltan2Interface_def.hpp:219
```

The offending line is:
https://github.com/trilinos/Trilinos/blob/master/packages/zoltan2/src/algorithms/partition/Zoltan2_AlgMultiJagged.hpp#L6787

**Comment by mhoemmen** [comment]
@jjellio I'm mainly just curious if this is an old version of the code.  This is something that changed very recently in master.  `git blame` says [17 days ago](https://github.com/trilinos/Trilinos/blame/ad4826a340f5e58766d81d2b531b12b8f4895ded/packages/tpetra/core/src/Tpetra_Map_def.hpp#L636), in fact.  Before then, `lgMapHost_` [was a CudaUVMSpace View](https://github.com/trilinos/Trilinos/blame/ad4826a340f5e58766d81d2b531b12b8f4895ded/packages/tpetra/core/src/Tpetra_Map_decl.hpp#L1295).  Now, it is a HostSpace View.

This matters mainly because there's no need for @kddevin to make these changes if the issue has already been fixed in a more robust way in `Tpetra::Map` itself.

**Comment by jjellio** [comment]
They are still sitting at pre-tempus API change (or atleast I am, and I think they still are) - we aren't on github trilinos either (I don't think any ATDM app is).

I view this as low priority, but it needs to be resolved at some point.  Right now, all of us EMPIRE fanbois are looking at pre-muelu setup, so we aren't immediately running into this (and it may be benign as well).

Why not leave this as WIP?  I hate that Karen put time into a patch that may not be needed. But on the flip side, understanding how unexpected MPI behavior can creep in is highly valuable.

**Comment by mhoemmen** [comment]
> Why not leave this as WIP? I hate that Karen put time into a patch that may not be needed. But on the flip side, understanding how unexpected MPI behavior can creep in is highly valuable.

Totally OK with me -- I just wanted to make sure people knew that this had been addressed in Tpetra::Map.  Thanks for working on this :-D

**Comment by rppawlo** [comment]
@kddevin - do we need this anymore? Should we close without merging?

**Comment by kddevin** [comment]
@william76 I'd like to keep this PR for a bit longer.

**Pull request closed without merging**